<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • ggnomics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Changelog" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ggnomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Overview.html">Overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/teunbrand/ggnomics/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/teunbrand/ggnomics/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="ggnomics-012" class="section level1">
<h1 class="page-header" data-toc-text="0.1.2">
<a href="#ggnomics-012" class="anchor"></a>ggnomics 0.1.2</h1>
<div id="major-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes" class="anchor"></a>Major changes</h2>
<ul>
<li>Started with a clean slate, removed previous ggnomics content.</li>
<li>Previous functionality unrelated to genomics now lives in the <a href="https://github.com/teunbrand/ggh4x">ggh4x</a> GitHub.</li>
</ul>
</div>
<div id="plotting" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting" class="anchor"></a>Plotting</h2>
<ul>
<li>Added fortify methods for <code><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">S4Vectors::DataFrame</a></code> and <code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">S4Vectors::Vector</a></code> classes, preserving S4 classes where possible.</li>
<li>Added <code>ggplot()</code> method for <code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">S4Vectors::Vector</a></code> class, which automatically adds <code><a href="../reference/coord_S4.html">coord_S4()</a></code>.</li>
</ul>
<div id="geoms" class="section level3">
<h3 class="hasAnchor">
<a href="#geoms" class="anchor"></a>Geoms</h3>
<ul>
<li>Added <code><a href="../reference/geom_range.html">geom_range()</a></code>, which is effectively <code>geom_rect()</code>, but can be parameterised by <code>IRanges::Ranges</code>-class for the x/y aesthetic instead of the xmin/xmax and ymin/ymax parameterisation.</li>
<li>Added <code><a href="../reference/geom_genemodel.html">geom_genemodel()</a></code> to display exon/intron structure of genes. Controlling the thickness of CDSs vs UTRs can be done using <code><a href="../reference/scale_exontype_discrete.html">scale_exontype_discrete()</a></code>.</li>
</ul>
</div>
<div id="stats" class="section level3">
<h3 class="hasAnchor">
<a href="#stats" class="anchor"></a>Stats</h3>
<ul>
<li>Added <code><a href="../reference/stat_coverage.html">stat_coverage()</a></code> to calculate how often a space is covered by a set of ranges. Treats <code>IRanges::Ranges</code>-class as closed-interval integers and doubles as open-interval.</li>
</ul>
</div>
<div id="positions" class="section level3">
<h3 class="hasAnchor">
<a href="#positions" class="anchor"></a>Positions</h3>
<ul>
<li>Added <code><a href="../reference/position_disjoint_ranges.html">position_disjoint_ranges()</a></code> to segregate overlapping ranges. Works similar to <code><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html">IRanges::disjointBins()</a></code>, but doesn’t require integer ranges.</li>
</ul>
</div>
<div id="scales" class="section level3">
<h3 class="hasAnchor">
<a href="#scales" class="anchor"></a>Scales</h3>
<ul>
<li>Added tree of ‘sister’ classes for ggplot’s discrete and continuous scales. It is the task of a scale to take S4 classes as input and wrap them inside an S3 representation and expose the correct training (range updating).</li>
<li>Added <code>scale_(x/y)_S4_continuous()</code> to handle continuous S4-class positions.</li>
<li>Added <code>scale_(x/y)_S4_discrete()</code> to handle discrete S4-class positions.</li>
<li>Added <code>scale_(colour/fill)_S4_continuous()</code> and <code>scale_(colour/fill)_S4_discrete()</code> to handle S4-class values.</li>
<li>Added <code>scale_(x/y)_genomic()</code>, which is almost identical to the <code>scale_(x/y)_S4_continuous()</code> scale except for labels, breaks and guide defaults.</li>
<li>Added <code><a href="../reference/guide_genomic_axis.html">guide_genomic_axis()</a></code> to display sequence name - position relation more clearly.</li>
<li>Dispatch rules for scale types; for <code><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html">S4Vectors::Factor</a></code> and <code><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">S4Vectors::Rle</a></code>, the levels/value slot is inspected to decide. Genomic classes in the <code>GenomicRanges</code> package go to <code>*_genomic</code> scales. Anything else is currently defaulting to the <code>*_S4_continuous</code> scales.</li>
<li>Added <code><a href="../reference/guide_ideogram_axis.html">guide_ideogram_axis()</a></code> to annotate positions with ideograms. Not implemented for genomic scales yet!</li>
</ul>
</div>
<div id="coordinates" class="section level3">
<h3 class="hasAnchor">
<a href="#coordinates" class="anchor"></a>Coordinates</h3>
<ul>
<li>Added <code><a href="../reference/coord_S4.html">coord_S4()</a></code> to control class specific final transformations. This coordinate system is added automatically when <code>ggplot()</code> is called on a <code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">S4Vectors::Vector</a></code> class. Should be added manually when S4 Vector data is added in, for example, layers. This allows the plot to ‘know’ that e.g. an <code><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges::IRanges</a></code> object underneath an <code>xmin</code> aesthetic should give the start, underneath <code>xmax</code> should give the end and under a plain <code>x</code> should give the middle.</li>
</ul>
</div>
</div>
<div id="internals" class="section level2">
<h2 class="hasAnchor">
<a href="#internals" class="anchor"></a>Internals</h2>
<ul>
<li>Added custom <code>ViewScale</code> ggprotos that can give minor labels and handle discontinuous limits (for e.g. <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GenomicRanges::GRanges</a></code>).</li>
<li>Added <code>forceflat</code> generic to map discontinuous classes to linear [0,1] interval. For genomic classes, mimicks <code><a href="https://rdrr.io/pkg/GenomicRanges/man/absoluteRanges.html">GenomicRanges::absoluteRanges()</a></code>; otherwise <code><a href="https://scales.r-lib.org//reference/rescale.html">scales::rescale()</a></code>.</li>
<li>Added range related generics for getting the range of a class, expanding a range and determining empty ranges.</li>
<li>Added plot arithmetic for WoodenHorse, so that Ranges obey e.g. nudging rules.</li>
<li>Added generic out-of-bounds variants for <code>censor()</code>, <code>squish()</code>, <code>squish_infinite()</code> and <code>discard()</code> from the scales package.</li>
<li>Added major/minor breaks <code>S4BreaksMajor()/S4BreaksMinor()</code> and labels <code><a href="../reference/S4LabelFormat.html">S4LabelFormat()</a></code> generics. They default to the scales package behaviour, but differ for genomic classes (e.g. GRanges).</li>
<li>Added <code>S4TransIdentity()</code> identity transformation that uses the major/minor breaks and label generics.</li>
<li>S3 wrapper-representation of S4 Vector class, so that they are compatible with data.frame operations. The S3 class is called <code>WoodenHorse</code> as they smuggle S4 Vectors into an S3 framework. Constructor is <code><a href="../reference/TheFallOfTroy.html">GreekSoldier()</a></code>, reversor is <code><a href="../reference/TheFallOfTroy.html">Nightfall()</a></code> and inspector of the underlying class in <code><a href="../reference/TheFallOfTroy.html">HelenOfTroy()</a></code>. There are two <code>WoodenHorse</code> variants; <code>BeechHorse</code> which carries the S4 Vector as a parallel attribute, and <code>OakHorse</code> which is an index into the S4 Vector attribute.</li>
<li>Added a <code>NEWS.md</code> file to track changes to the package.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Teun van den Brand.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


